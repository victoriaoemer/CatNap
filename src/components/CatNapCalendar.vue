<template>
  <div class="flex flex-col gap-1 h-max p-2 rounded-2xl text-white w-max">
    <div class="grid grid-cols-7 gap-1 font-black text-md text-center">
      <div
        v-for="(day, index) in weekDays"
        :key="day"
        :class="['p-1 rounded', { 'bg-purple text-white': index === currentDay }]"
      >
        {{ day }}
      </div>
    </div>
    <div class="grid grid-cols-7 gap-1 justify-items-center">
      <div
        v-for="empty in startEmptyDays"
        :key="'empty-' + empty"
        class="calendar-cell empty"
      ></div>

      <div class="calendar-cell" v-for="day in daysInMonthCount" :key="'day-' + day">
        <div
          :class="[
            'rounded-full',
            emotionClassForDay(day) === 'bg-gray-400' ? 'w-1 h-1' : 'w-2 h-2',
            emotionClassForDay(day),
          ]"
        ></div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
const weekDays = ['M', 'T', 'W', 'T', 'F', 'S', 'S']

const today = new Date()
const currentDay = (today.getDay() + 6) % 7 // monday = 0, sunday = 6

const currentMonth = today.getMonth()
const currentYear = today.getFullYear()

const firstDayOfMonth = new Date(currentYear, currentMonth, 1).getDay()
const startDayIndex = (firstDayOfMonth + 6) % 7
const daysInMonthCount = new Date(currentYear, currentMonth + 1, 0).getDate()
const startEmptyDays = Array.from({ length: startDayIndex }, (_, i) => i + 1)

const emotionClassForDay = (day: number) => {
  if (props.userData.data === undefined) {
    return 'bg-gray-400'
  }

  const dateKey = `${String(day).padStart(2, '0')}/${String(currentMonth + 1).padStart(2, '0')}/${currentYear}`
  const dayData = props.userData.data[dateKey]

  if (dayData && dayData.emotion) {
    const emotionMap = {
      1: 'bg-emotion-red',
      2: 'bg-emotion-orange',
      3: 'bg-emotion-green',
    }
    return emotionMap[dayData.emotion as keyof typeof emotionMap] || 'bg-gray-400'
  }
  return 'bg-gray-400'
}

const props = defineProps({
  userData: {
    type: Object,
    required: true,
    default: () => ({ data: {} }),
  },
})
</script>

<style scoped>
.calendar-cell {
  width: 1rem;
  height: 1rem;
  display: flex;
  align-items: center;
  justify-content: center;
}

.calendar-cell.empty {
  background-color: transparent;
}
</style>
