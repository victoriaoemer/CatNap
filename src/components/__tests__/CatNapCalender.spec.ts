import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import CatNapCalendar from '@/components/CatNapCalendar.vue';

describe('CatNapCalendar.vue', () => {
  it('renders weekdays correctly', () => {
    const wrapper = mount(CatNapCalendar, {
      props: {
        userData: { data: {} },
      },
    });

    const weekDays = wrapper.findAll('.grid.grid-cols-7 > .p-1.rounded'); // pecifically target weekday header cells
    expect(weekDays.length).toBe(7);
    const expectedDays = ['M', 'T', 'W', 'T', 'F', 'S', 'S'];
    weekDays.forEach((day, index) => {
      expect(day.text().trim()).toBe(expectedDays[index]);
    });
  });

  it('renders the correct number of days and empty cells', () => {
    const wrapper = mount(CatNapCalendar, {
      props: {
        userData: { data: {} },
      },
    });

    const emptyCells = wrapper.findAll('.calendar-cell.empty');
    const daysInMonth = new Date(new Date().getFullYear(), new Date().getMonth() + 1, 0).getDate();
    const dayCells = wrapper.findAll('.calendar-cell:not(.empty)');

    expect(emptyCells.length).toBeGreaterThanOrEqual(0); // Dynamic based on start day
    expect(dayCells.length).toBe(daysInMonth);
  });

  it('applies the correct emotion classes for days', () => {
    const userData = {
      data: {
        '01/01/2025': { emotion: 1 },
        '02/01/2025': { emotion: 2 },
        '03/01/2025': { emotion: 3 },
      },
    };

    const wrapper = mount(CatNapCalendar, {
      props: {
        userData,
      },
    });

    const dayCells = wrapper.findAll('.calendar-cell');

    const firstDay = dayCells.find((cell) => cell.text().trim() === '1');
    const secondDay = dayCells.find((cell) => cell.text().trim() === '2');
    const thirdDay = dayCells.find((cell) => cell.text().trim() === '3');

    if (firstDay) {
      expect(firstDay.find('.rounded-full').classes()).toContain('bg-emotion-red');
    }
    if (secondDay) {
      expect(secondDay.find('.rounded-full').classes()).toContain('bg-emotion-orange');
    }
    if (thirdDay) {
      expect(thirdDay.find('.rounded-full').classes()).toContain('bg-emotion-green');
    }
  });

  it('matches the snapshot', () => {
    const wrapper = mount(CatNapCalendar, {
      props: {
        userData: { data: {} },
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
