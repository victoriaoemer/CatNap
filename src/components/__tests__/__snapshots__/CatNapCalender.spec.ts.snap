// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CatNapCalendar.vue > matches the snapshot 1`] = `
"<div data-v-d61c7321="" class="flex flex-col gap-1 h-max p-2 rounded-2xl text-white w-max">
  <div data-v-d61c7321="" class="grid grid-cols-7 gap-1 font-black text-md text-center">
    <div data-v-d61c7321="" class="p-1 rounded">M</div>
    <div data-v-d61c7321="" class="p-1 rounded">T</div>
    <div data-v-d61c7321="" class="p-1 rounded">W</div>
    <div data-v-d61c7321="" class="p-1 rounded">T</div>
    <div data-v-d61c7321="" class="p-1 rounded bg-purple text-white">F</div>
    <div data-v-d61c7321="" class="p-1 rounded">S</div>
    <div data-v-d61c7321="" class="p-1 rounded">S</div>
  </div>
  <div data-v-d61c7321="" class="grid grid-cols-7 gap-1 justify-items-center">
    <div data-v-d61c7321="" class="calendar-cell empty"></div>
    <div data-v-d61c7321="" class="calendar-cell empty"></div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
    <div data-v-d61c7321="" class="calendar-cell">
      <div data-v-d61c7321="" class="rounded-full w-1 h-1 bg-gray-400"></div>
    </div>
  </div>
</div>"
`;
